introspection-engine/connectors/sql-introspection-connector/src/calculate_datamodel.rs

```diff
@@ -1,28 +1,80 @@
 use crate::commenting_out_guardrails::commenting_out_guardrails;
 use crate::misc_helpers::*;
+use crate::misc_helpers::{is_prisma_1_or_11_list_table, is_relay_table};
 use crate::sanitize_datamodel_names::sanitize_datamodel_names;
 use crate::SqlIntrospectionResult;
 use datamodel::{dml, Datamodel, FieldType, Model};
-use introspection_connector::IntrospectionResult;
+use introspection_connector::{IntrospectionResult, Version};
+use quaint::connector::SqlFamily;
 use sql_schema_describer::*;
 use tracing::debug;
 
 /// Calculate a data model from a database schema.
-pub fn calculate_model(schema: &SqlSchema) -> SqlIntrospectionResult<IntrospectionResult> {
+pub fn calculate_datamodel(schema: &SqlSchema, family: &SqlFamily) -> SqlIntrospectionResult<IntrospectionResult> {
     debug!("Calculating data model.");
 
+    // Detect the version of the database
+    let migration_table = schema.tables.iter().any(|table| is_migration_table(&table));
+    let has_prisma_1_join_table = schema.tables.iter().any(|table| is_prisma_1_point_0_join_table(&table));
+    let has_prisma_1_1_or_2_join_table = schema
+        .tables
+        .iter()
+        .any(|table| is_prisma_1_point_1_or_2_join_table(&table));
+    let mut uses_on_delete = false;
+    let mut always_has_created_at_updated_at = true;
+    let mut uses_non_prisma_types = false;
+    let mut has_inline_relations = false;
```

Comment:
The state space of this function is getting pretty big :p

But it's hard to think of a good way to structure this without doing too much repetitive works (lots of looping) or introducing as much or more state.

Reply:
Yeah, I also think it is pretty ugly but could not think of a good way to pull the whole functionality in a separate place without handing around references everywhere.

Reply:
Reworked it after our discussion